test/pkg/integration/workspace.go

```diff
@@ -352,32 +342,28 @@ func (it *Test) WaitForWorkspaceStop(instanceID string) (lastStatus *wsmanapi.Wo
 				continue
 			}
 
-			if status.Conditions.Failed != "" {
-				it.t.Errorf("workspace instance %s failed: %s", instanceID, status.Conditions.Failed)
-				return
-			}
-			if status.Phase == wsmanapi.WorkspacePhase_STOPPED {
-				lastStatus = status
+			if check(it.t, status) {
 				return
 			}
 		}
 	}()
 
 	// maybe the workspace has started in the meantime and we've missed the update
-	desc, _ := wsman.DescribeWorkspace(it.ctx, &wsmanapi.DescribeWorkspaceRequest{Id: instanceID})
-	if desc != nil {
-		switch desc.Status.Phase {
-		case wsmanapi.WorkspacePhase_STOPPED:
-			return desc.Status
-		}
+	desc, err := wsman.DescribeWorkspace(it.ctx, &wsmanapi.DescribeWorkspaceRequest{Id: instanceID})
+	if err != nil {
+		it.t.Fatalf("cannot get workspace: %q", err)
+		return
+	}
+	if check(it.t, desc.Status) {
+		close(done)
```

Comment:
This might race the `defer close()` in the Go routine. If we need to close the same chanel here, we should sync it using `sync.Once` or better yet, use a separate channel.